### PR TITLE
WIP: Topic/better

### PR DESCRIPTION
**Pull request by ShaneDelmore**

### CODE REVIEW

File: scalafix/rules/src/main/scala/rsc/rules/RscCompat.scala [hunk 1]
```diff
@@ -11,39 +11,39 @@ import rsc.rules.syntax._
 import scala.meta._
```
Comment (xeno-by) [inline scalafix/rules/src/main/scala/rsc/rules/RscCompat.scala:11]: Please split the pull request into several commits: 1) update to v2, 2) add unit tests, 3) fix the broken integration test. Current arrangement makes it hard to review these changes in isolation.

File: scalafix/rules/src/main/scala/rsc/rules/RscCompat.scala [hunk 2]
```diff
@@ -11,39 +11,39 @@ import rsc.rules.syntax._
...
-  override def fix(ctx: RuleCtx): Patch = {
-    val targets = collectRewriteTargets(ctx)
-    targets.map(ascribeReturnType(ctx, _)).asPatch
+  override def fix(implicit doc: SemanticDocument): Patch = {
+    val targets = collectRewriteTargets(doc)
+    targets.map(ascribeReturnType(doc, _)).asPatch
   }
 
-  private case class RewriteTarget(
+  case class RewriteTarget(
```
Comment (xeno-by) [inline scalafix/rules/src/main/scala/rsc/rules/RscCompat.scala:37]: Why are this class and the methods below changed to public?

File: scalafix/rules/src/main/scala/rsc/rules/RscCompat.scala [hunk 3]
```diff
@@ -11,39 +11,39 @@ import rsc.rules.syntax._
...
       env: Env,
       before: Token,
       name: Name,
       after: Token,
       body: Term,
       parens: Boolean)
 
-  private def collectRewriteTargets(ctx: RuleCtx): List[RewriteTarget] = {
+  def collectRewriteTargets(doc: SemanticDocument): List[RewriteTarget] = {
+    implicit val impDoc = doc
```
Comment (xeno-by) [inline scalafix/rules/src/main/scala/rsc/rules/RscCompat.scala:46]: Looks like the `doc` parameter should be implicit after the migration to v1. A dummy implicit val looks non-idiomatic.

File: scalafix/rules/src/main/scala/rsc/rules/RscCompat.scala [hunk 4]
```diff
@@ -115,18 +115,21 @@ case class RscCompat(legacyIndex: SemanticdbIndex, config: RscCompatConfig)
...
       val returnTypeString = {
-        val symbol = target.name.symbol.get.syntax
+        val symbol = target.name.symbol.value
         config.hardcoded.get(symbol) match {
           case Some(returnTypeString) =>
             returnTypeString
-          case _ =>
+          case _ if index.symbols.get(symbol).isDefined =>
+            //FIXME, ignore symbols that are not found as pattern syms are returning as localn
+            //with no symbo on lookup due to https://github.com/scalameta/scalameta/issues/1725
```
Comment (xeno-by) [inline scalafix/rules/src/main/scala/rsc/rules/RscCompat.scala:132]: In Rsc master, FIXME comments have a special format: `// FIXME: <link to a ticket>` and TODO comments are prohibited. This: 1) makes sure that issues are only tracked in the issue tracker, 2) makes code more concise since most of the time a naked link to a ticket is sufficient to convey meaning.

Please change the new FIXME comments to accommodate this convention (todo items can be migrated to issues at https://github.com/ShaneDelmore/rsc). If you'd like, feel free to keep the existing TODO comments unchanged for the time being to simplify this pull request.

File: scalafix/rules/src/main/scala/rsc/rules/RscCompat.scala [hunk 5]
```diff
@@ -151,12 +154,15 @@ case class RscCompat(legacyIndex: SemanticdbIndex, config: RscCompatConfig)
                 printer.pprint(returnType)
                 printer.toString
             }
+          case _ =>
+            //FIXME: Do nothing until https://github.com/scalameta/scalameta/issues/1725 is resolved
```
Comment (xeno-by) [inline scalafix/rules/src/main/scala/rsc/rules/RscCompat.scala:158]: Ditto.

File: scalafix/rules/src/main/scala/rsc/rules/pretty/SemanticdbPrinter.scala [hunk 6]
```diff
@@ -13,7 +13,8 @@ import scala.meta.internal.{semanticdb => s}
 import scala.meta.internal.semanticdb.Scala._
 import scala.meta.internal.semanticdb.Scala.{Descriptor => d}
 import scala.meta.internal.semanticdb.Scala.{Names => n}
-import scalafix.internal.v0._
+import scala.meta.internal.semanticdb.{Scala, Signature, TypeRef, TypeSignature}
```
Comment (xeno-by) [inline scalafix/rules/src/main/scala/rsc/rules/pretty/SemanticdbPrinter.scala:16]: The idiomatic way of referring to `Signature` etc is to `import scala.meta.internal.{semanticdb => s}` (already done above) and then say `s.Signature` etc.

File: scalafix/rules/src/main/scala/rsc/rules/pretty/SemanticdbPrinter.scala [hunk 7]
```diff
@@ -81,6 +82,43 @@ class SemanticdbPrinter(env: Env, index: DocumentIndex, config: RscCompatConfig)
...
+  }
+
+  def isSameOrAlias(sym: String, envSym: String) =
+    (sym == envSym) || dealias(envSym).contains(sym)
+
+  def dealias(sym: String): Option[String] = {
+    index.symbols.get(sym).flatMap(s => symbolOf(s.signature))
+  }
+
+  def extractName(sym: String): Option[Scala.Names.Name] = {
```
Comment (xeno-by) [inline scalafix/rules/src/main/scala/rsc/rules/pretty/SemanticdbPrinter.scala:109]: The idiomatic way of referring to `Name` is `import scala.meta.internal.semanticdb.Scala.{Names => n}` and then `n.Name`.

File: scalafix/rules/src/main/scala/rsc/rules/pretty/SemanticdbPrinter.scala [hunk 8]
```diff
@@ -81,6 +82,43 @@ class SemanticdbPrinter(env: Env, index: DocumentIndex, config: RscCompatConfig)
...
+    // The logic to improve on this is left for future work.
+    sym.desc match {
+      case d.Term(value) => Some(n.TermName(value))
+      case d.Type(value) => Some(n.TypeName(value))
+      case d.Package(value) => Some(n.TermName(value))
+      case d.Parameter(value) => Some(n.TermName(value))
+      case d.TypeParameter(value) => Some(n.TypeName(value))
+      case other => None
+    }
+  }
```
Comment (xeno-by) [inline scalafix/rules/src/main/scala/rsc/rules/pretty/SemanticdbPrinter.scala:120]: These helper methods look unrelated to the printer. Please move them into `rsc.rules.semantics`. For example, `extractName` could become an extension method `nameOpt` in `SymbolOps`.

File: scalafix/rules/src/main/scala/rsc/rules/pretty/SemanticdbPrinter.scala [hunk 9]
```diff
@@ -81,6 +82,43 @@ class SemanticdbPrinter(env: Env, index: DocumentIndex, config: RscCompatConfig)
     case _ => sys.error(s"unsupported tree $tree")
   }
 
+  def symbolOf(signature: Signature): Option[String] =
```
Comment (xeno-by) [inline scalafix/rules/src/main/scala/rsc/rules/pretty/SemanticdbPrinter.scala:85]: This method needs to either: 1) have a better name, or 2) get inlined into isSameOrAlias.

File: scalafix/rules/src/main/scala/rsc/rules/pretty/SemanticdbPrinter.scala [hunk 10]
```diff
@@ -81,6 +82,43 @@ class SemanticdbPrinter(env: Env, index: DocumentIndex, config: RscCompatConfig)
...
+    signature match {
+      case TypeSignature(_, lower: TypeRef, upper: TypeRef) if lower == upper => Some(lower.symbol)
```
Comment (xeno-by) [inline scalafix/rules/src/main/scala/rsc/rules/pretty/SemanticdbPrinter.scala:87]: The idiomatic names for these variables are `lo` and `hi`.

File: scalafix/tests/src/test/scala/rsc/rules/pretty/SemanticdbPrinterTest.scala [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+  private val doc = TestHelper.syntacticDoc(
+    tree.pos.input,
+    LazyValue.now(tree),
+    DiffDisable.empty,
+    scalafixConfig
+  )
+  private val semanticDoc = TestHelper.semanticDoc(doc, aTest.semanticdbPath, classLoader, symtab)
+  private val index = DocumentIndex(semanticDoc)
+  private val printer = new SemanticdbPrinter(predefEnv, index, RscCompatConfig.default)
+  import printer._
```
Comment (xeno-by) [inline scalafix/tests/src/test/scala/rsc/rules/pretty/SemanticdbPrinterTest.scala:44]: It looks like the unit tests below don't use any of the fields above except `symtab`. Given that, I don't think we should add 40+ lines of tricky initialization code tied to Scalafix internals.

Comment (xeno-by) [inline scalafix/tests/src/test/scala/rsc/rules/pretty/SemanticdbPrinterTest.scala:44]: See https://github.com/xeno-by/rsc/commit/ed0f19cf985eb1d883ef6bb0a66b7d0d76b5cfff for a proposed simplification of `DocumentIndex`, after which it can take a `SymbolTable` and nothing else.

File: scalafix/tests/src/test/scala/rsc/rules/pretty/SemanticdbPrinterTest.scala [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+
+  //Testing is very light for now but this is enough to demonstrate calling syntax
+  //and exercise the test harness
```
Comment (xeno-by) [inline scalafix/tests/src/test/scala/rsc/rules/pretty/SemanticdbPrinterTest.scala:47]: Looks like a temporary comment for PR reviewers rather than a permanent comment. If that's the case, note taken and the comment may be deleted now.

File: scalafix/rules/src/main/scala/rsc/rules/pretty/SemanticdbPrinter.scala [hunk 13]
```diff
@@ -81,6 +82,43 @@ class SemanticdbPrinter(env: Env, index: DocumentIndex, config: RscCompatConfig)
...
+    extractName(sym)
+      .map(env.lookup) match {
+      // TODO: If the lookup returns NoSymbol, we can insert an import and skip the prefix.
+      // The logic to implement this is left for future work.
+      case None => false
+      case Some(symFromScope) => isSameOrAlias(sym, symFromScope)
+    }
+  }
+
+  def isSameOrAlias(sym: String, envSym: String) =
```
Comment (xeno-by) [inline scalafix/rules/src/main/scala/rsc/rules/pretty/SemanticdbPrinter.scala:102]: What does `envSym` mean?

File: scalafix/rules/src/main/scala/rsc/rules/pretty/SemanticdbPrinter.scala [hunk 14]
```diff
@@ -130,9 +156,10 @@ class SemanticdbPrinter(env: Env, index: DocumentIndex, config: RscCompatConfig)
             rep("[", args, ", ", "]")(normal)
           }
         case s.SingleType(pre, sym) =>
-          // TODO: Also check for config.better.
-          val prettyPre = if (pre == s.NoType) sym.trivialPrefix(env) else pre
-          opt(prettyPre, ".")(prefix)
+          if (!config.better) {
```
Comment (xeno-by) [inline scalafix/rules/src/main/scala/rsc/rules/pretty/SemanticdbPrinter.scala:159]: This logic is incorrect because it doesn't check whether `sym` is in scope. If it isn't, the prefix cannot be omitted.

File: scalafix/rules/src/main/scala/rsc/rules/RscCompat.scala (hunk 1 above)
Comment (xeno-by) [inline scalafix/rules/src/main/scala/rsc/rules/RscCompat.scala:11]: Also, please submit future pull requests against `topic/better` in my repository. This will avoid extraneous commits in the "Commits" tab.

**Review by xeno-by — CHANGES_REQUESTED** [review]
Thank you for the pull request! I left a bunch of comments. In a nutshell: 1) v1 migration looks good except for a minor thing, 2) the fix to the broken test is incorrect, 3) unit tests can be simplified significantly.

File: scalafix/rules/src/main/scala/rsc/rules/RscCompat.scala (hunk 1 above)
Comment (ShaneDelmore) [reply scalafix/rules/src/main/scala/rsc/rules/RscCompat.scala:11]: I've gotten used to phablet style squashing everything down. I will split them up tomorrow.

File: scalafix/rules/src/main/scala/rsc/rules/RscCompat.scala (hunk 2 above)
Comment (ShaneDelmore) [reply scalafix/rules/src/main/scala/rsc/rules/RscCompat.scala:37]: The class needs to be public to work with the v1 config loader. The others were public so I could test them more easily. I created adhoc tests for them to get the rule working then removed the tests as they were mostly for my learning and not otherwise useful.

File: scalafix/rules/src/main/scala/rsc/rules/pretty/SemanticdbPrinter.scala (hunk 6 above)
Comment (ShaneDelmore) [reply scalafix/rules/src/main/scala/rsc/rules/pretty/SemanticdbPrinter.scala:16]: Thanks. This is IJ auto-imports. I forgot about your renaming imports strategy but will revise.

File: scalafix/tests/src/test/scala/rsc/rules/pretty/SemanticdbPrinterTest.scala (hunk 11 above)
Comment (ShaneDelmore) [reply scalafix/tests/src/test/scala/rsc/rules/pretty/SemanticdbPrinterTest.scala:44]: Nice

File: scalafix/rules/src/main/scala/rsc/rules/pretty/SemanticdbPrinter.scala (hunk 13 above)
Comment (ShaneDelmore) [reply scalafix/rules/src/main/scala/rsc/rules/pretty/SemanticdbPrinter.scala:102]: Could just as well have been sum' or sym2, but it means the sum pulled from env.lookup. Suggestions for a better name?

File: scalafix/rules/src/main/scala/rsc/rules/pretty/SemanticdbPrinter.scala (hunk 14 above)
Comment (ShaneDelmore) [reply scalafix/rules/src/main/scala/rsc/rules/pretty/SemanticdbPrinter.scala:159]: I originally checked if sym was in scope but it ended up printing one extra layer of name, as in collection.mutable.map instead of mutable.map. Will need to investigate further it seems.

File: scalafix/rules/src/main/scala/rsc/rules/RscCompat.scala (hunk 1 above)
Comment (ShaneDelmore) [reply scalafix/rules/src/main/scala/rsc/rules/RscCompat.scala:11]: I wondered why that happened, thanks for pointing out I PR'd against master accidentally.

File: scalafix/rules/src/main/scala/rsc/rules/RscCompat.scala (hunk 2 above)
Comment (xeno-by) [reply scalafix/rules/src/main/scala/rsc/rules/RscCompat.scala:37]: > The class needs to be public to work with the v1 config loader. 

Why?

File: scalafix/rules/src/main/scala/rsc/rules/pretty/SemanticdbPrinter.scala (hunk 13 above)
Comment (xeno-by) [reply scalafix/rules/src/main/scala/rsc/rules/pretty/SemanticdbPrinter.scala:102]: `sym2` doesn't work well in my opinion, because `sym` and `sym2` are not interchangeable - only `sym2` is dealiased, but not `sym`.

Comment (xeno-by) [reply scalafix/rules/src/main/scala/rsc/rules/pretty/SemanticdbPrinter.scala:102]: I can't come up with a better name at the moment. In situations like that, I inline helpers into their callsites, which saves the mental effort of naming.

File: scalafix/rules/src/main/scala/rsc/rules/RscCompat.scala (hunk 2 above)
Comment (ShaneDelmore) [reply scalafix/rules/src/main/scala/rsc/rules/RscCompat.scala:37]: Make it private and it fails to compile with a inaccessible error. I will reproduce later and add the message here.

File: scalafix/rules/src/main/scala/rsc/rules/pretty/SemanticdbPrinter.scala (hunk 13 above)
Comment (ShaneDelmore) [reply scalafix/rules/src/main/scala/rsc/rules/pretty/SemanticdbPrinter.scala:102]: I prefer not to inline here because I lose the benefit of naming, isSameOrAlias. I will add a doc comment for clarity.

File: scalafix/tests/src/test/scala/rsc/rules/pretty/SemanticdbPrinterTest.scala (hunk 11 above)
Comment (ShaneDelmore) [reply scalafix/tests/src/test/scala/rsc/rules/pretty/SemanticdbPrinterTest.scala:44]: I just ran out of time last nite. I needed all of that setup to create the printer tests I want to add.
